ffmpeg/ffmpeg_test.go

```diff
@@ -602,6 +604,159 @@ func TestTranscoder_MuxerOpts(t *testing.T) {
+					switch outAccel {
+						case Nvidia:
+							switch outCodec {
+								case VP8, VP9:
+									continue
+								}
+					}
+					opts = append(opts, TranscodeOptionsTest{
+						InputCodec:   inCodec,
+						OutputCodec:  outCodec,
+						InputAccel:   inAccel,
+						OutputAccel:  outAccel,
+						Profile:      prof,
+						QualityCheck: true,
+					})
+				}
+			}
+		}
+	}
+	return opts
+}
+
+func CodecsComboTest(t *testing.T, options []TranscodeOptionsTest) {
+	run, dir := setupTest(t)
+	defer os.RemoveAll(dir)
+	sampleName := dir + "/test.ts"
+	var inName, outName, qName string
+	cmd := `
+    # set up initial input; truncate test.ts file
+    ffmpeg -loglevel warning -i "$1"/../transcoder/test.ts -c:a copy -c:v copy -t 1 test.ts
+  `
+	run(cmd)
+	var err error
+	for i := range options {
+		curOptions := options[i]
+		switch curOptions.InputCodec {
+			case VP8, VP9:
+				inName = dir + "/test_in.mkv"
+			case H264, H265:
+				inName = dir + "/test_in.ts"
+		}
+		switch curOptions.OutputCodec {
+			case VP8, VP9:
+				outName = dir + "/out.mkv"
+				qName = dir + "/sw.mkv"
+			case H264, H265:
+				outName = dir + "/out.ts"
+				qName = dir + "/sw.ts"
+		}
+		// if non-h264 test requested, transcode to target input codec first
+		prepare := true
+		if curOptions.InputCodec != H264 {
+			profile := P720p60fps16x9
+			profile.Encoder = curOptions.InputCodec
+			err = Transcode2(&TranscodeOptionsIn{
+				Fname: sampleName,
+				Accel: Software,
+			}, []TranscodeOptions{
+				{
+					Oname:   inName,
+					Profile: profile,
+					Accel:   Software,
+				},
+			})
+			if err != nil {
+				t.Error(err)
+				t.Fail()
+				prepare = false
+			}
+		} else {
+			inName = sampleName
+		}
+		transcode := true
+		targetProfile := curOptions.Profile
+		targetProfile.Encoder = curOptions.OutputCodec
+		err = Transcode2(&TranscodeOptionsIn{
+			Fname: inName,
+			Accel: curOptions.InputAccel,
+		}, []TranscodeOptions{
+			{
+				Oname:   outName,
+				Profile: targetProfile,
+				Accel:   curOptions.OutputAccel,
+			},
+		})
+		if err != nil {
+			t.Error(err)
+			t.Fail()
+			transcode = false
+		}
+		quality := prepare && transcode
+		if quality && curOptions.QualityCheck {
+			// software transcode for image quality check
+			err = Transcode2(&TranscodeOptionsIn{
+				Fname: inName,
+				Accel: Software,
+			}, []TranscodeOptions{
+				{
+					Oname:   qName,
+					Profile: targetProfile,
+					Accel:   Software,
+				},
+			})
+			if err != nil {
+				t.Error(err)
+				t.Fail()
+				quality = false
+			}
+			cmd = fmt.Sprintf(`
+    # compare using ssim and generate stats file
+    ffmpeg -loglevel warning -i %s -i %s -lavfi '[0:v][1:v]ssim=stats.log' -f null -
+    # check image quality; ensure that no more than 5 frames have ssim < 0.95
+    grep -Po 'All:\K\d+.\d+' stats.log | awk '{ if ($1 < 0.95) count=count+1 } END{ exit count > 5 }'
+  `, outName, qName)
+			if quality {
+				quality = run(cmd)
+				if !quality {
+					t.Fail()
```

Comment:
Shouldn't the cmd trigger a `t.Error()` if it returns a non-zero exit code?

Reply:
I think the func produced by `setupTest` never checks exit code, it will trigger error if [something was written to stderr](https://github.com/livepeer/lpms/blob/5c05cbbeb53bf7cfbc413c5b33ddadf3d3699719/ffmpeg/ffmpeg_test.go#L33), which won't be the case here.

Reply:
Fixed.